### ISSUE TITLE
ESLint Plugin React Hooks throws false negative / forces bad design

### ISSUE THREAD

**Issue by janhesters** [facebook/react#15417]
**Do you want to request a *feature* or report a *bug*?**
*feature*

**What is the current behavior?**
**TLDR:** If you use `useMemo` and pass a function to a child component from its parent component, `eslint-plugin-react-hooks` will yell at you if you leave out that function in the dependencies. But since it's a new function in every render, adding it to the `useMemo` dependencies defeats the purpose of `useMemo`.

Here is my case where I encountered the problem. I'm building an app that renders a table of contacts. All contacts are editable on the fly, which is why every cell of this table is an input.

Here is the code for the container component that calls its styled presentational component:

```js
function ContactCell({
  cell: { name },
  value: original,
  searchValue,
  changeContact,
  contactId,
  removeContactChange
}) {
  const [value, setValue] = useState(original);
  function handleChange(e) {
    setValue(e.target.value);
    const change = { id: contactId, value: e.target.value, key: name };
    if (e.target.value === original) {
      removeContactChange(change);
    } else {
      changeContact(change);
    }
  }
  const matched =
    searchValue && original.toLowerCase().includes(searchValue.toLowerCase());
  const component = useMemo(
    () => (
      <Styled
        name={name}
        onChange={handleChange}
        value={value}
        matched={matched}
      />
    ),
    // in this line the plugin complains that handleChange is not a dependency.
    [name, value, matched]
  );
  return <>{component}</>;
}
```

In another part of the app users can search for contacts using a `searchValue`. I use `useMemo` here to avoid all cells being re-rendered whenever the `searchValue` changes and only re-render if they are matched or the value changes.

Now if I were to include `handleChange`, all contact cells would re-render if the `searchValue` changes, since `handleChange` is new in every render.

**What is the expected behavior?**
Maybe the linter plugin recognizes functions declared in the scope of the component and allows you to leave them out?

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**
React 16.8.6 in Google Chrome on MacOS

**Comment by gaearon** [comment]
It is a mistake to leave our functions. They can capture values from state and props — and then you’ll run into more confusing bugs. 

The correct fixes are described here:

https://reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies

TLDR:

* Either move your function inside the Hook. 
* Or move it outside of the component. 
* If neither is feasible, wrap it in `useCallback` and specify *its* dependencies. Then it won’t change except the cases where it *should* change.

**Comment by gaearon** [comment]
(If you’re in a mood for a longread, I also wrote about it here: https://overreacted.io/a-complete-guide-to-useeffect/#but-i-cant-put-this-function-inside-an-effect)

**Comment by janhesters** [comment]
@gaearon Thank you for your answer. I have read both the docs (back when hooks came out) and your article (back when you published it) and I didn't remember these fixes.

Thank you for reminding and teaching me! 🙏

**Comment by janhesters** [comment]
@gaearon 

I have one more follow up question. What about catching functions exposed by a custom hook?

```js
function useFocus(initialValue = false) {
  const [focused, setFocused] = useState(initialValue);

  function handleFocus() {
    setFocused(true);
  }

  function handleBlur() {
    setFocused(false);
  }

  return { focused, onBlur: handleBlur, onFocus: handleFocus };
}
```

If I leave out `onBlur` and `onFocus` the linter plugin complains. But adding them, makes the component re-render too often. And the plugin also prevents you from using custom hooks in `useMemo`.

Again, thank you for your help 🙏

**Comment by gaearon** [comment]
Same fix:

```js
function useFocus(initialValue = false) {
  const [focused, setFocused] = useState(initialValue);

  const handleFocus = useCallback(() => {
    setFocused(true);
  }, []);

  const handleBlur = useCallback(() => {
    setFocused(false);
  }, []);

  return { focused, onBlur: handleBlur, onFocus: handleFocus };
}
```

**Comment by janhesters** [comment]
@gaearon Thank you so much!
